### PR TITLE
Catch up to the block height of Agora immediately after Stoa is executed

### PR DESCRIPTION
**Pull request by MichaelKim20**
If a new block is not pushed after the Stoa is run, it will not be recovered to the height of the block that Agora has.
To solve this, I added the ability to catch up to the block height of Agora using the block height value obtained from Agora when the Stoa was executed.

Relates to #140 

The text below is an execution log of the Stoa.
I made blocks up to height 2 in Agora.
And I ran Stoa.
Stoa caught up to height 2 blocks
And the new block height 3 and 4 were also pushed, and normally recorded in the database.

```
[Stoa] 2020-11-11T05:02:45.893Z info Agora endpoint: http://127.0.0.1:4002
[Stoa] 2020-11-11T05:02:45.896Z info sqlite3 database filename: /stoa/data/main.db
[Stoa] 2020-11-11T05:02:45.914Z info Connected to Agora, block height is 2
[Stoa] 2020-11-11T05:02:45.919Z info Listening to requests on: 0.0.0.0:2828
[Stoa] 2020-11-11T05:02:45.961Z info Recovered a block with block height of 0
[Stoa] 2020-11-11T05:02:45.972Z info Recovered a block with block height of 1
[Stoa] 2020-11-11T05:02:45.982Z info Recovered a block with block height of 2
[Stoa] 2020-11-11T05:02:54.256Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x58bc869bee498afc, distance : 8
[Stoa] 2020-11-11T05:03:04.290Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x58bc869bee498afc, distance : 8
[Stoa] 2020-11-11T05:03:14.210Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x58bc869bee498afc, distance : 8
[Stoa] 2020-11-11T05:03:22.587Z info Saved a block with block height of 3
[Stoa] 2020-11-11T05:03:24.225Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x85bb716531de74af, distance : 9
[Stoa] 2020-11-11T05:03:24.458Z info Saved a pre-image enroll_key : 0x8c1561a4475df42a, hash : 0xff9abdcd5472690e, distance : 9
[Stoa] 2020-11-11T05:03:24.676Z info Saved a pre-image enroll_key : 0x46883e83778481d6, hash : 0xd50b71784cad3883, distance : 9
[Stoa] 2020-11-11T05:03:24.927Z info Saved a pre-image enroll_key : 0x94908ec79866cf54, hash : 0xc931cf38ce39ca36, distance : 9
[Stoa] 2020-11-11T05:03:25.031Z info Saved a pre-image enroll_key : 0xdb3931bd87d2cea0, hash : 0x3e5a8d2bd620df6e, distance : 9
[Stoa] 2020-11-11T05:03:25.035Z info Saved a pre-image enroll_key : 0x4dde806d2e09367f, hash : 0xb55beb9070b59b6e, distance : 9
[Stoa] 2020-11-11T05:03:34.236Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x85bb716531de74af, distance : 9
[Stoa] 2020-11-11T05:03:44.186Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x85bb716531de74af, distance : 9
[Stoa] 2020-11-11T05:03:54.195Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0x85bb716531de74af, distance : 9
[Stoa] 2020-11-11T05:04:04.123Z info Saved a block with block height of 4
[Stoa] 2020-11-11T05:04:04.192Z info Saved a pre-image enroll_key : 0xb20da9cfbda971f3, hash : 0xc5bc12b9cdcdb605, distance : 10
```

### CODE REVIEW

File: src/Stoa.ts [hunk 1]
```diff
@@ -100,17 +100,26 @@ class Stoa extends WebService
...
+                            .catch((err: any) => {
+                                reject(err);
+                            })
+                    },
+                    (err) => {
+                        logger.error(`Error: Could not connect to Agora node: ${err.toString()}`);
+                        process.exit(1);
+                    }
+                );
+        });
```
Comment (Geod24) [inline src/Stoa.ts]: You do not need to wrap this in a `Promise`. It's a redundant pattern which I eliminated in many places in the code base, and we should be careful not to re-introduce it.
```suggestion
        return this.agora.getBlockHeight()
                .then(
                    (height) => {
                        logger.info(`Connected to Agora, block height is ${height.toString()}`);
                        return super.start();
                    })
                    .then(() => {
                        this.pending = this.pending.then(() => { return this.catchup(height); });
                    }),
                    (err) => {
                        logger.error(`Error: Could not connect to Agora node: ${err.toString()}`);
                        process.exit(1);
                    }
                );
        });
```

Promises can be chained. `a.then((result) { return result.action(); })` will chain the `Promise` returned by `action` to the `Promise` returned by `a`. This is much easier to write and read than nesting things, and requires a single catch.
Not that `throw`ing inside a promise is equivalent to `reject`ing it.

Comment (Geod24) [reply src/Stoa.ts]: Note ^ The above code wasn't tested, but the idea is simple: chain, do not nest.

File: src/Stoa.ts [hunk 2]
```diff
@@ -366,7 +373,7 @@ class Stoa extends WebService
             (async () => {
                 try
                 {
-                    let max_blocks : bigint = height.value - expected_height.value;
+                    let max_blocks : bigint = height.value - expected_height.value + ((block == null) ? 1n : 0n);
```
Comment (MichaelKim20) [inline src/Stoa.ts:376]: Add 1 because the block is null and must be recovered to height during catch-up.

File: src/Stoa.ts [hunk 3]
```diff
@@ -396,8 +403,11 @@ class Stoa extends WebService
                     // Save a block just received
                     if (height.value <= expected_height.value)
                     {
-                        await this.ledger_storage.putBlocks(Block.reviver("", block));
-                        logger.info(`Saved a block with block height of ${height.toString()}`);
+                        if (block != null)
```
Comment (MichaelKim20) [inline src/Stoa.ts:406]: During catch-up, skip this process because the block is null.

File: src/Stoa.ts (hunk 1 above)
Comment (MichaelKim20) [reply src/Stoa.ts]: Well, I think it's good to use 'then' duplicated.
I accepted the opinion, fixed the code, and made some modifications through the test.

File: src/Stoa.ts (hunk 2 above)
Comment (AndrejMitrovic) [reply src/Stoa.ts:376]: It's better to add this comment in the code instead of in github.

**Review by AndrejMitrovic — APPROVED** [review]

File: src/Stoa.ts [hunk 4]
```diff
@@ -491,6 +501,42 @@ class Stoa extends WebService
         });
     }
 
+    /**
+     * Catches up to block height of Agora
+     * This is done only once immediately after Stoa is executed.
+     * @param height The block height of Agora
+     * @private
```
Comment (TrustHenry) [inline src/Stoa.ts]: Don't need it.

**Review by TrustHenry — APPROVED** [review]
LGTM

**Pull request merged**